external-crates/move/crates/move-compiler/src/typing/syntax_methods.rs

```diff
@@ -0,0 +1,234 @@
+            "This index function expects {} parameters",
+            index_finfo.signature.parameters.len()
+        );
+        let index_mut_msg = format!(
+            "This mutable index function expects {} parameters ",
+            mut_finfo.signature.parameters.len()
+        );
+        let mut diag = diag!(
+            TypeSafety::IncompatibleSyntaxMethods,
+            (index.loc, index_msg),
+            (index_mut.loc, index_mut_msg),
+        );
+        diag.add_note("Index operations on the same type must take the name number of parameters");
+        context.env.add_diag(diag);
+        return false;
+    }
+
+    // Now we simply want to skolomize the types and make sure they are the same. To do this, we
+    // instantiate the type variables and ground them both to new type parameters. Finally, we walk
+    // both types and make sure they agree. Note that  we discard any constraints introduced here
+    // because those will be checked later when the index functions are actually used.
+
+    let prev_constraints = std::mem::take(&mut context.constraints);
+
+    let mut valid = true;
+
+    let mut subst = core::Subst::empty();
+
+    let index_ty = core::make_function_type(context, index_ann_loc, index_module, index_fn, None);
+    let index_mut_ty =
+        core::make_function_type(context, index_ann_loc, index_mut_module, index_mut_fn, None);
+
+    for (ndx, (index_tyarg, index_mut_tyarg)) in index_ty
+        .ty_args
+        .iter()
+        .zip(index_mut_ty.ty_args.iter())
+        .enumerate()
+    {
+        if let Ok((subst_, _)) = core::join(subst.clone(), index_tyarg, index_mut_tyarg) {
+            subst = subst_;
+        } else {
+            context
+                .env
+                .add_diag(ice!((index.loc, "Failed in validating syntax methods")));
+            valid = false;
+        }
+        // This essentially fakes skolomeziation so that we can fail when the types aren't the same.
+        let id = N::TParamID::next();
+        let user_specified_name = sp(index_tyarg.loc, format!("_{}", ndx).into());
+        let tparam = N::TParam {
+            id,
+            user_specified_name,
+            abilities: AbilitySet::all(index_ann_loc),
+        };
+        if let Ok((subst_, _)) = core::join(
+            subst.clone(),
+            index_tyarg,
+            &sp(index_tyarg.loc, N::Type_::Param(tparam)),
+        ) {
+            subst = subst_;
+        } else {
+            context
+                .env
+                .add_diag(ice!((index.loc, "Failed in validating syntax methods")));
+            valid = false;
+        }
+    }
+
+    fn ty_str(ty: &N::Type) -> String {
+        core::error_format(ty, &core::Subst::empty())
+    }
+
+    fn ty_str_(ty: &N::Type_) -> String {
+        core::error_format_(ty, &core::Subst::empty())
+    }
+
+    if let Ok((subst_, _)) = core::subtype(
+        subst.clone(),
+        &index_mut_ty.params[0].1,
+        &index_ty.params[0].1,
```

Comment:
To avoid some of this index math, maybe zip+enumerate them ahead of time, and then pop off the front for this first case